src/scheduler/MpiWorld.cpp

```diff
@@ -4,28 +4,87 @@
 #include <faabric/util/func.h>
 #include <faabric/util/gids.h>
 #include <faabric/util/macros.h>
+#include <faabric/util/testing.h>
 
-/* Each MPI rank runs in a separate thread, thus we use TLS to maintain the
- * per-rank data structures.
- */
+// Each MPI rank runs in a separate thread, thus we use TLS to maintain the
+// per-rank data structures
 static thread_local std::vector<
   std::unique_ptr<faabric::transport::MpiMessageEndpoint>>
   mpiMessageEndpoints;
+
 static thread_local std::vector<
   std::shared_ptr<faabric::scheduler::MpiMessageBuffer>>
   unackedMessageBuffers;
+
 static thread_local std::set<int> iSendRequests;
+
 static thread_local std::map<int, std::pair<int, int>> reqIdToRanks;
 
+// These long-lived sockets are used by each world to communicate rank-to-host
+// mappings. They are thread-local to ensure separation between concurrent
+// worlds executing on the same host
+static thread_local std::unique_ptr<
+  faabric::transport::AsyncRecvMessageEndpoint>
+  ranksRecvEndpoint;
```

Comment:
Yes I went back and forth on these endpoints, they can't be ephemeral otherwise we have into issues running locally when the sender might get destructed before the receiver and it hangs. For this reason they have to be linked to a scope higher in the stack, but they can't be tied to the world itself as the thread that creates it might not be the one that ends up interacting with the socket. Therefore they ended up living in this TLS.